### ISSUE TITLE
Optimize peak memory for flash _scaled_dot_product_attention_math

### ISSUE THREAD

**Issue by 1274085042** [pytorch/pytorch#139612]
### 🚀 The feature, motivation and pitch

  

Currently, `_scaled_dot_product_attention_math` uses `at::_safe_softmax(attn, -1)`, which increases memory usage compared to the previous `at::softmax(attn, -1)` implementation. We found that in some large model applications, the peak memory exceeds the memory capacity due to `_safe_softmax`, resulting in OOM.
The peak memory usage of `_scaled_dot_product_attention_math` is as follows:

**PT2.1**
![image](https://github.com/user-attachments/assets/02d99554-26ee-4979-87c5-a6abedf302de)

pt=2.1.0a0+git7bcf7da: used=2360.94MB  

**PT2.6**  
![image](https://github.com/user-attachments/assets/428fa869-49f4-4884-ad71-6bd37eec669f)  
pt=2.6.0a0+gitf4ee5a2: used=3898.94MB

Test Scripts：  
```python
import torch
from torch.nn.attention import sdpa_kernel, SDPBackend
from torch.profiler import profile, ProfilerActivity

device = "cuda"

torch.backends.cuda.allow_fp16_bf16_reduction_math_sdp(True)
size = (1,32, 4096, 128)
q = torch.randn(size, device=device, dtype=torch.half, requires_grad=True)
k = torch.randn(size, device=device, dtype=torch.half, requires_grad=True)
v = torch.randn(size, device=device, dtype=torch.half, requires_grad=True)
with sdpa_kernel(SDPBackend.MATH):
    torch.cuda.memory._record_memory_history()
    res=torch.nn.functional.scaled_dot_product_attention(q, k, v)
    torch.cuda.memory._dump_snapshot("./sdpa_math_fwd_where_2.6.pickle")
    free, total = map(lambda x: x/2**20, torch.cuda.mem_get_info())
    used=total-free
    print(f'pt={torch.__version__}: {used=:0.2f}MB, {free=:0.2f}MB, {total=:0.2f}MB')
```

@drisspg @albanD 

### 

### Alternatives

I want to implement `_safe_softmax` in the following way, `at::where_out` reduces peak memory usage by reusing  memory
```cpp
Tensor _safe_softmax(
    const Tensor& self,
    int64_t dim,
    std::optional<ScalarType> dtype) {
  auto out = at::softmax(self, dim, dtype);
  const auto neg_inf = at::scalar_tensor(-std::numeric_limits<float>::infinity(), at::TensorOptions().dtype(out.dtype()).device(out.device()));
  const auto masked = self.eq(neg_inf);
  const auto masked_rows = all(masked, dim, true);
  const auto zero = at::scalar_tensor(0.0, at::TensorOptions().dtype(out.dtype()).device(out.device()));
  // reuse storage for out
 return at::where_out(out, masked_rows, zero, out);
}
```

The peak memory usage of `_scaled_dot_product_attention_math` is as follows:
![image](https://github.com/user-attachments/assets/ec0c74c6-06c4-4722-9b1b-cbc824d5f05e)

pt=2.6.0a0+gitf4ee5a2: used=2874.94MB

cc @ptrblck @msaroufim @drisspg @mikaylagawarecki

**Comment by janeyx99** [comment]
cc @drisspg
